### ISSUE TITLE
Alternative Deferred howto

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6180]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#6180|
|Type|enhancement|
|Created|2012-11-09 17:05:54Z|
|Branch|https://github.com/twisted/twisted/tree/alternative-deferred-6180-3|

I've written a document based on a talk I gave years ago. I think it might be appropriate for inclusion in core Twisted documentation.

The document can be found here: https://gist.github.com/4040065 

Next steps are to make a branch & submit for review properly, but I'd welcome feedback in the mean time.

Attachments:

* [proposed.rewords.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/722/722944f89091ce5d9b1c5fbdfd00568555f67a8a/0fb3ac264b46ef6853f30b9d717f715e4dbbc577.patch) (3043 bytes) - added by teratorn on 2013-01-19 21:28:51Z - proposed rewording of various paragraphs
* [deferred-conclusions.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/722/722944f89091ce5d9b1c5fbdfd00568555f67a8a/433163225abfe814675b7e679f1c5fb5c4e5b1ad.patch) (2283 bytes) - added by teratorn on 2013-02-02 21:39:42Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6180 6180
type__enhancement enhancement
reporter__jml jml
priority__normal normal
milestone__ 
branch__branches_alternative_deferred_6180_3 branches/alternative-deferred-6180-3
branch_author__jml jml
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1352480754000000 1352480754000000
changetime__1360065085000000 1360065085000000
version__None None
owner__jml jml
cc__teratorn
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_1"></a>|@radix commented|
|-|-|

I really like the document you wrote, but I don't think it'd be a good idea to create an "Alternative" Deferred howto. I'm not trying to provide stop energy, though. I just think your document should just replace the existing howto after it's been checked for "feature parity" or whatever.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_2"></a>|@radix commented|
|-|-|

[1] By the way, I noticed a typo. In your explanation of addBoth, you say it's equivalent to d.addCallback(g, g), which should have an "s" as in "addCallbacks".

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml commented|
|-|-|

Thanks, I'm glad you like the doc. I've fixed up that typo in my local copy.

I've done basically zero thinking on how it should integrate with the rest of the documentation beyond what you just said: that it should do so rather than be chucked in. 

I like the idea of expanding it to match the existing doc for feature parity.  Another idea might be to have two docs, with this one being the intro and the second being a reference.  Or one doc being "using Deferreds" and the other one being "writing Deferred-returning code". Hmm. I'll try a couple of things & see what I like.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_4"></a>|@jml commented|
|-|-|

(In [36564]) Branching to 'alternative-deferred-6180'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml commented|
|-|-|

I've added the document as "Introduction to Deferred", have converted to Lore, and added an introduction and conclusion, as per the writing standard.

I tried doing a pass over the existing deferred document, but my eyes are glazing over. I'd very much appreciate it if someone (the reviewer perhaps?) could identify the bits in that doc that aren't necessary given the addition of the introduction.

I've used `em` as the way to mark up the first use of a technical term.  I'm not sure if that's correct.

Feedback welcomed!

jml

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml commented|
|-|-|

(In [36583]) Recommended changes from Chipaca (refs [#6180](https://github.com/twisted/twisted/issues/6180)). Thanks!

http://dpaste.org/PWOZE/

**Comment by twisted-trac** [comment]
|<img alt="lewq's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|lewq commented|
|-|-|

I think the document is great, it would definitely have helped me when I first learned about Deferreds.  A few minor niggles:

Line 29:
```
  assumes no knowledge of Twisted.  Understanding what an asynchronous will
```

Perhaps this should be "an asynchronous function"?  Or if it's really not necessary to understand what an async function is, perhaps you could drop the whole "Understanding..." sentence?

'em' seems like a fine tag to use for initial definitions of things.

```
          Why <q>roughly</q>?  Because if <code class="python">f</code>
	  raises, <code class="python">g</code> will be passed a <code class="API"
	  base="twisted.python.failure">Failure</code> object representing the
	  exception.  Otherwise, <code class="python">g</code> will be passed the
	  asynchronous equivalent of the return value
	  of <code class="python">f()</code> (i.e. <code class="python">y</code>).
```

This bit is hard to parse at first glance because of the large number of single letter variable names.  It's not obvious why 'y' is equivalent to 'f()' until you realise it's on the LHS of the assignment operator, so maybe drop the brackets.  Personally I preferred the contrived examples to the single letter variables, but either way.

It's not clear from the comments above or the branch whether the doc is intended to replace the Deferred doc (defer.xhtml).  I do think that it's better than the original (clearer and more engaging), but it doesn't cover failure.trap, nor does it have the diagrams which some might find helpful, so I'd propose perhaps keeping both, and making the new (simpler one) the default linked to from the Twisted website's documentation page, so long as the new one links to the old one for "further reading".

Cheers,
Luke

**Comment by twisted-trac** [comment]
|<img alt="lewq's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|lewq set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_10"></a>|@teratorn commented|
|-|-|

Hi jml,

Would like to drop in with a review also, if I may. I like the new intro document very much! Overall it strikes me as more approachable than the defer.html howto, especially for beginners or those new to async programming. I've been over the text several times, trying out structural modifications. I've attached a patch that, I think, makes significant structural improvements to the flow of the text, without changing your core messages much, if at all.

Let me know what you think - I mostly just tried to convey your ideas in simpler terms, perhaps more explicitly - and with slightly less, in my mind, "superfluous prose". Also I like the technique of taking an explict contrived example, then transitioning to generic variables. I think it helps study the different error handling forms without distraction. Cheers, and thanks again, -teratorn

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_11"></a>|@jml commented|
|-|-|

(In [36934]) Branching to 'alternative-deferred-6180-2'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_12"></a>|@jml commented|
|-|-|

(In [36935]) Branching to 'alternative-deferred-6180-3'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_13"></a>|@jml commented|
|-|-|

(In [36939]) Remove unnecessary, poorly-formed sentence, per lewq's advice

* refs [#6180](https://github.com/twisted/twisted/issues/6180)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_14"></a>|@jml commented|
|-|-|

(In [36940]) Factor in most of teratorn's suggestions.

* refs [#6180](https://github.com/twisted/twisted/issues/6180)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @teratorn|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_16"></a>|@teratorn set owner to @jml|
|-|-|

OK checked over the diff - looks pretty good to me. It's just missing a newsfile. Please merge after adding one! :)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_17"></a>|@wallrj commented|
|-|-|

I get this when I view it directly in Firefox.

 * http://twistedmatrix.com/trac/export/36948/branches/alternative-deferred-6180-3/doc/core/howto/defer-intro.xhtml

Do you need the entity if the encoding is UTF-8?

```
XML Parsing Error: undefined entity
Location: http://twistedmatrix.com/trac/export/36948/branches/alternative-deferred-6180-3/doc/core/howto/defer-intro.xhtml
Line Number 18, Column 6:  yet&emdash;that's why you are here! 
-----^
```

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_18"></a>|@teratorn commented|
|-|-|

Dang, for some reason I thought that was right without checking the lore processed output :(
I guess it is supposed to be "mdash" not "emdash" ?

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_19"></a>|@jml commented|
|-|-|

(In [37046]) Fix incorrect HTML markup for dash (refs [#6180](https://github.com/twisted/twisted/issues/6180))

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_20"></a>|@jml set owner to @teratorn|
|-|-|

rwall, you need to browse the HTML generated by lore, rather than the XHTML document.

teratorn, fixed.  I've also run the lore build locally and verified that the output has no errors and the doc looks good.

The TwistedDevelopment page says how:
```
cd ./path/to/Twisted-checkout/doc; ../bin/lore/lore --config template=core/howto/template.tpl
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_21"></a>|@jml commented|
|-|-|

(In [37047]) Add NEWS file (refs [#6180](https://github.com/twisted/twisted/issues/6180))

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_22"></a>|@teratorn set owner to @jml|
|-|-|

Thanks jml.  I've been through the doc once more and fixed two typos in the branch. Also the conclusion, though there is nothing actually *wrong* with it - it seemed like a bit bland to me, and I thought it could be worded in a way to sound a bit more, "active", perhaps? Patch attached, just "sounds better" to my ear - let me know what you think :)

+1 to merge either way.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_23"></a>|@jml commented|
|-|-|

I've applied the patch on the final paragraph.  I prefer the "You have been ..."  bit as is, even though it's not as active, because it permits the bullet point items to be more actively phrased.  Thanks very much for the suggestions though, and the typo fixes.

Merging now.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_24"></a>|@jml set status to `closed`|
|-|-|

(In [37070]) Introductory howto on using Deferreds

Author: jml

Reviewer: teratorn

Fixes: [#6180](https://github.com/twisted/twisted/issues/6180)

Document explaining how Deferreds work and how to map basic Python
control flow to Deferreds.  The existing Deferred howto is now linked
as a reference document.
